src/python/pants/init/specs_calculator.py

```diff
@@ -69,10 +71,37 @@ def calculate_specs(
             "The `--changed-*` options are only available if Git is used for the repository."
         )
 
-    changed_files = tuple(changed_options.changed_files(maybe_git_worktree.git_worktree))
+    (files_with_text_blocks,) = session.product_request(
+        FilesWithTextBlocks, [Params(bootstrap_environment)]
+    )
+    changed_files = tuple(
+        file
+        for file in changed_options.changed_files(maybe_git_worktree.git_worktree)
+        # We want to exclude the file from the normal processing flow if it has associated
+        # targets with text blocks. These files are handled with special logic.
+        if file not in files_with_text_blocks
+    )
     file_literal_specs = tuple(FileLiteralSpec(f) for f in changed_files)
 
-    changed_request = ChangedRequest(changed_files, changed_options.dependents)
+    text_blocks = FrozenDict(
+        (
+            path,
+            # Hunk stores information about the old block and the new block.
+            # Here we only care about the final state, so we take `hunk.right`.
+            TextBlocks(hunk.right for hunk in hunks),
+        )
+        for path, hunks in changed_options.diff_hunks(
+            maybe_git_worktree.git_worktree,
+            files_with_text_blocks,
+        ).items()
+    )
+    logger.debug("changed text blocks: %s", text_blocks)
+
+    changed_request = ChangedRequest(
+        sources=changed_files,
+        dependents=changed_options.dependents,
+        text_blocks=text_blocks,
+    )
```

Comment:
> If something goes wrong because of the optimizations I don't want the whole pr to be reverted

Gotcha.

We can indeed postpone optimizations here as follow up.